File: frontend/Sources/Domain/Collection.swift
import Foundation

package struct Collection: Sendable, Equatable, Codable, Identifiable {
    package let id: UUID
    package var name: String
    package var description: String?
    package var icon: String?
    package var parentId: UUID?
    package var sortOrder: Int
    package var filterJson: CollectionFilter?
    package var isSmart: Bool
    package let createdAt: Date
    package let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, name, description, icon
        case parentId = "parent_id"
        case sortOrder = "sort_order"
        case filterJson = "filter_json"
        case isSmart = "is_smart"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

package struct CollectionFilter: Sendable, Equatable, Codable {
    package var query: String?
    package var fileType: String?
    package var tags: [String]?

    package init(query: String? = nil, fileType: String? = nil, tags: [String]? = nil) {
        self.query = query
        self.fileType = fileType
        self.tags = tags
    }

    enum CodingKeys: String, CodingKey {
        case query
        case fileType = "file_type"
        case tags
    }
}

package struct CollectionListResponse: Sendable, Codable {
    package let collections: [Collection]
    package let total: Int
    package let limit: Int
    package let offset: Int
}

package struct TagListResponse: Sendable, Codable {
    package let tags: [String]
}

File: backend/src/cortex/domain/collection.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from uuid import UUID, uuid4


@dataclass
class Collection:
    id: UUID
    name: str
    description: str | None = None
    icon: str | None = None
    parent_id: UUID | None = None
    sort_order: int = 0
    filter_json: dict | None = None
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))

    @property
    def is_smart(self) -> bool:
        return self.filter_json is not None

    @staticmethod
    def new(
        name: str,
        description: str | None = None,
        icon: str | None = None,
        parent_id: UUID | None = None,
        sort_order: int = 0,
        filter_json: dict | None = None,
    ) -> Collection:
        return Collection(
            id=uuid4(),
            name=name,
            description=description,
            icon=icon,
            parent_id=parent_id,
            sort_order=sort_order,
            filter_json=filter_json,
        )
